components/search_results/search_results.tsx

```diff
@@ -0,0 +1,296 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+/* eslint-disable react/no-string-refs */
+
+import React, {useEffect, useRef} from 'react';
+import Scrollbars from 'react-custom-scrollbars';
+
+import classNames from 'classnames';
+
+import {debounce} from 'mattermost-redux/actions/helpers';
+
+import * as Utils from 'utils/utils.jsx';
+import {searchHintOptions} from 'utils/constants';
+
+import SearchResultsHeader from 'components/search_results_header';
+import SearchResultsItem from 'components/search_results_item';
+import SearchHint from 'components/search_hint/search_hint';
+import LoadingSpinner from 'components/widgets/loading/loading_wrapper';
+import NoResultsIndicator from 'components/no_results_indicator/no_results_indicator';
+import FlagIcon from 'components/widgets/icons/flag_icon';
+
+import {NoResultsVariant} from 'components/no_results_indicator/types';
+
+import type {Props} from './types';
+
+const GET_MORE_BUFFER = 30;
+
+const renderView = (props: Record<string, unknown>): JSX.Element => (
+    <div
+        {...props}
+        className='scrollbar--view'
+    />
+);
+
+const renderThumbHorizontal = (props: Record<string, unknown>): JSX.Element => (
+    <div
+        {...props}
+        className='scrollbar--horizontal'
+    />
+);
+
+const renderThumbVertical = (props: Record<string, unknown>): JSX.Element => (
+    <div
+        {...props}
+        className='scrollbar--vertical'
+    />
+);
+
+export const shouldRenderFromProps = (props: Props, nextProps: Props): boolean => {
+    // Shallow compare for all props except 'results'
+    for (const key in nextProps) {
+        if (!Object.prototype.hasOwnProperty.call(nextProps, key) || key === 'results') {
+            continue;
+        }
+
+        if (nextProps[key] !== props[key]) {
+            return true;
+        }
+    }
+
+    // Here we do a slightly deeper compare on 'results' because it is frequently a new
+    // array but without any actual changes
+    const {results} = props;
+    const {results: nextResults} = nextProps;
+
+    if (results.length !== nextResults.length) {
+        return true;
+    }
+
+    for (let i = 0; i < results.length; i++) {
+        // Only need a shallow compare on each post
+        if (results[i] !== nextResults[i]) {
+            return true;
+        }
+    }
+
+    return false;
+};
+
+interface NoResultsProps {
+    variant: NoResultsVariant,
+    titleValues?: Record<string, JSX.Element|string>
+    subtitleValues?: Record<string, JSX.Element|string>
```

Comment:
These types can be `Record<string, React.ReactNode>`

Reply:
done